ui/gtk/src/main.rs

```diff
@@ -67,10 +72,227 @@ impl Into<i32> for Column {
+
+                    for table in arch.tables.iter() {
+                        for entry in table.entries.iter() {
+                            let float_len = entry.file_size as f32;
+
+                            let formatted_size = match binary_prefix(float_len) {
+                                Standalone(bytes) => format!("{} B", bytes),
+                                Prefixed(prefix, n) => format!("{:.2} {}B", n, prefix),
+                            };
+
+                            ei_store.insert_with_values(None,
+                                &[
+                                    Column::Name.into(),
+                                    Column::Type.into(),
+                                    Column::Size.into(),
+                                    Column::Offset.into(),
+                                ],
+                                &[
+                                    &entry.file_id.to_string(),
+                                    &table.header.file_extension(),
+                                    &formatted_size,
+                                    &format!("{:#X}", entry.file_offset),
+                                ]
+                            );
+                        }
+                    }
+
+                    archive.replace(Some(arch));
+                }
+            }
+        }
+    });
+}
+
+fn enable_extract_button(archive: Rc<Cell<Option<Archive>>>,
+                         extract_button: Button,
+                         entryinfo_tree: TreeView) {
+    extract_button.connect_clicked(move |_this| {
+        if let Some(dest_dir_path) = select_dir_dialog("Select a directory to extract to",
+                                                       WindowType::Toplevel,
+                                                       gtk::FileChooserAction::SelectFolder) {
+            if let Some(arch) = archive.take() {
+                let sel = entryinfo_tree.get_selection();
+                let (mut sel_paths, model) = sel.get_selected_rows();
+
+                if sel_paths.len() == 0 {
+                    sel.select_all();
+                    let (s, _) = sel.get_selected_rows();
+                    sel_paths = s;
+                    sel.unselect_all();
+                }
+
+                for sel_path in sel_paths {
+                    if let Some(iter) = model.get_iter(&sel_path) {
+                        let val = model.get_value(&iter, 0);
+                        let name = val
+                            .get::<String>()
+                            .expect(&format!("Unable to convert gtk::Type::String {:?} to a Rust String", val));
+
+                        for table in arch.tables.iter() {
+                            if let Some(data) = table.find_file_contents(name.parse::<u32>().unwrap()) {
+                                let mut output_filepath = dest_dir_path.clone();
+                                output_filepath.push(name.replace("\\", "/"));
```

Comment:
Please append the extension (determined by the table type) to this filepath.

Currently entry `156` (for example) is written to disk as `156` but what the user is going to want is `156.slp`.